### ISSUE TITLE
feat(agent): expose bounded selected-token semantic reads from native inference

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#13432]
# feat(agent): expose bounded selected-token semantic reads from native inference

```routing
lane: agent
paths: ["internal/agent/semantic_read.go", "internal/agent/semantic_read_test.go"]
expected_steps: 7
```

## Working spine
State plus bounded typed questions -> native prompt-final logits -> normalized candidate distributions and provenance receipt, with zero generated answer tokens.

## Current state
The native planner already computes prompt-final logits and chosen-token log probabilities, but no bounded typed-decision surface accepts caller-declared candidate labels and returns their normalized distribution. Consumers must generate and parse text or reach into engine internals.

## Why this is next
A first-class semantic read is reusable engine capability and permits honest comparison between direct-logit decisions and generated structured output. It is also the smallest public seam required to test Jev-style reads on supported native backends.

## Parent context
Parent #11572 owns the Strix Halo coding-agent native-performance campaign. This bounded primitive gives that campaign a zero-generated-token decision path to measure. Independent open implementations such as SemIf demonstrate the interface pattern; this ticket extends existing native planner seams and does not vendor their runtimes or branding.

## Core through-line
Add an internal agent request/result contract for atomic choice/binary/ordinal questions over a shared state. Tokenize declared labels, reject ambiguous multi-token or duplicate encodings in v1, read the native prompt-final distribution without decoding text, normalize only declared candidates, and bind model/tokenizer/revision plus timing to the receipt.

## Gold-plating boundary
No HTTP API stabilization, SDK, policy engine, routing decision, calibration claim, training, model download, OpenJev compatibility promise, dense-logit exposure, or private factory integration. This leaf proves the native engine primitive only.

## Done condition / witness
Focused tests prove candidate distributions come from the same raw logits as the native planner, are schema-bounded and deterministic, emit zero generated tokens, and fail closed on unsupported encodings/backends.

Witness: `go test -v -count=1 ./internal/agent -run '^TestSemanticRead_'`

## Verifiable Witness
`go test -v -count=1 ./internal/agent -run '^TestSemanticRead_'`

## Definition of done
- [ ] Choice, binary, and ordinal questions share one state and return only declared candidate IDs.
- [ ] Candidate probabilities are derived from native logits and sum to one within a declared tolerance.
- [ ] Unsupported multi-token/duplicate label encodings fail closed with typed reasons.
- [ ] Receipt binds model, tokenizer, source revision, prompt digest, latency, and zero generated tokens.
- [ ] Existing generation, cache, and sampling behavior remains byte-compatible when the feature is unused.
- [ ] Tests compare against an independent softmax oracle and cover cancellation/non-finite logits.

## Acceptance gate
The native semantic read is invoked in focused tests through the real planner seam; presence-only structs or mocked probabilities do not satisfy the ticket.

## Closure binding
Close from a signed public commit limited to the declared paths, with `Fixes #<issue>` and `(fak agent)`.

## Lane
agent

## Likely files
- internal/agent/semantic_read.go
- internal/agent/semantic_read_test.go

## Expected steps
7

## Value
- Centrality: Core
- P1 Context: advanced - one shared state supports several bounded decisions
- P2 Net value: advanced - avoids generated answer tokens and parsing
- P3 Adaptation: preserved - caller-defined candidates remain runtime inputs
- P4 Operations: preserved - primitive exposes no authority or side effect

## Work estimate
Estimate: 7 points

## Overall completion contribution
Contribution: 7/24 points

## Completion standard
development

**Comment by anthony-chaudhary** [comment]
# feat(agent): implement the bounded selected-token semantic read over native prompt-final logits

## Working spine
State plus bounded typed questions -> native prompt-final logits -> normalized candidate distributions and a provenance receipt, with zero generated answer tokens.

## Core through-line
Add an internal `agent` request/result contract for atomic choice/binary/ordinal questions over a shared state. Validate that each declared label is one exact round-trip token, reject ambiguous multi-token or duplicate encodings, read the native prompt-final distribution without decoding text, normalize only declared candidates, and bind model/tokenizer/revision plus timing to the receipt.

## Current state

`rg -n "SemanticRead|semantic_read" --glob '*.go'` in the public `fak` tree returns **no files**:

```
internal/agent/semantic_read.go       -> ABSENT
internal/agent/semantic_read_test.go  -> ABSENT
```

Issue fak#13432 is OPEN. The private consumers gated on it (#2423 bakeoff, #2424 intake advisor)
are CLOSED and green. The private half of the Jev-style semantic-decision track is landed; the
public engine primitive is not.

The name is now resolved, which is what unblocks this. "Jev" was recorded `ABSENT` in two prior
notes because repo-local search found nothing. It is **TypeSafe's closed semantic-decision
service** (eval paper **arXiv:2609.24965**, *"Jev for Scientific Decisions"*, Boyuan Deng et al.),
reproduced openly as **SemIf** (formerly **OpenJev**), **MIT**,
`github.com/TheoLeeCJ/SemIf` @ `1f2dea3e25379f9dfc98cb83c324f00ab5deda37` (default branch is
`master`; a `main` probe 404s).

The mechanism is verified at source (`src/semif_phase1/direct.py`, 77 lines), not from a README:
prompt-final last-position logits restricted to declared answer slots, `softmax` over **only**
those candidates, and **zero answer tokens generated**.

The seam already exists; only the projection is missing. `internal/agent/inkernel_decode.go`
already computes prompt-final logits and the chosen-token log-probability
(`lp := float64(logits[token]) - maxLogit - math.Log(expSum)`), and `inkernel_reuse.go` caches
prompt-final logits. A semantic read is therefore a **bounded projection over an already-computed
vector** — no new prefill cost, no new decode path.

## Scope

In scope: an internal `internal/agent` request/result contract for atomic choice / binary /
ordinal questions over one shared state, reading native prompt-final logits and normalizing only
caller-declared candidates.

Out of scope (gold-plating boundary): do **not** vendor SemIf or its runtimes or branding, and
make no Jev/OpenJev compatibility promise — port the *validation discipline* only. No HTTP API
stabilization, SDK, policy engine, routing decision, calibration claim, model download, dense-logit
exposure, or private factory integration in this leaf.

## Witness

`rg -n "SemanticRead|semantic_read" --glob '*.go'` in the public tree returns no files today; after this lands, the named tests must execute non-vacuously through the real planner seam.

## Done condition / witness

Witness: `go test -v -count=1 ./internal/agent -run '^TestSemanticRead_'` in the public repo,
invoking the real planner seam. Presence-only structs or mocked probabilities do not satisfy it.

## Definition of done

- [ ] `SemanticRead(state, questions[])` returns only caller-declared candidate IDs for choice / binary / ordinal questions over one shared state.
- [ ] Candidate probabilities derive from the same raw logits as the native planner and sum to 1 within a declared tolerance, compared against an independent softmax oracle.
- [ ] **Fail closed** (typed reasons, not panics) on: a label that is not one exact round-trip token; duplicate/colliding slot encodings; answer-boundary tokenization drift; fewer than two finite candidate scores; non-finite or cancelled logits.
- [ ] Receipt binds model, tokenizer, source revision, prompt digest, latency, and asserts **zero generated tokens**.
- [ ] Existing generation, cache, and sampling behavior is byte-compatible when the feature is unused.

## Prior art (pinned, license-verified; `api.github.com` was 403 and was not used)

| Project | HEAD | License | gfx1151 | Class |
|---|---|---|---|---|
| `TheoLeeCJ/SemIf` | `1f2dea3e2537` | MIT | plausible | DIRECT-PORT (pattern only) |
| `ggml-org/llama.cpp` | `f46bc30cb6a7` | MIT | **yes** (`__gfx1151__` in `ggml-cuda/vendors/hip.h`) | ADAPT |
| `vllm-project/vllm` | `25965ee583b1` | Apache-2.0 | **yes** (`platforms/rocm.py`: `0x1586 -> AMD_Radeon_8060S`) | ADAPT (`prompt_logprobs` shape) |
| `guidance-ai/guidance` | `21b1d90dfbeb` | MIT | via llama.cpp | ADAPT (`select`/`get_logits` analogue) |
| `dottxt-ai/outlines` | `c52af8472c6f` | Apache-2.0 | via llama.cpp/vLLM | ADAPT (masks, does not normalize) |
| `sgl-project/sglang` | `542c817bfae2` | Apache-2.0 | **none found** | INSPIRE-ONLY |

Outlines / SGLang / LMQL **constrain then sample a token**; SemIf **reads and emits none**. Same
word, opposite mechanism and cost model.

## Adjudication note (private side, this pass)

`TestSemanticAdvisorDeterminism` was red on the private trunk: `semanticAdvisorReceiptDigest`
cleared `Digest` but not the wall-clock `GeneratedAt` it is computed after, so identical corpora
hashed differently. Fixed by excluding `generated_at` from the content digest; the full
`SemanticAdvisor` suite is green. This does not change the criteria above; it removes a
determinism defect in the consumer-side receipt this primitive will feed.
